app/src/main/java/rak/pixellwp/cycling/models/Palette.kt

```diff
@@ -10,31 +10,31 @@ import kotlin.math.min
 class Palette(val id: String = "", colors: List<Int>, val cycles: List<Cycle>) {
     constructor(id: String = "", paletteJson: PaletteJson) : this(id, paletteJson.parsedColors, paletteJson.cycles)
 
-    val baseColors = colors.toMutableList()
-    var colors = baseColors.toMutableList()
+    val baseColors = colors.toIntArray()
+    val colors = baseColors.clone()
 
     fun blendPalette(previous: Palette, next: Palette, percent: Int): Palette {
-        val mixedColors = baseColors.toIntArray()
-        for (i in 0 until baseColors.size){
+        val mixedColors = baseColors.clone()
+        for (i in baseColors.indices){
             mixedColors[i] = (fadeColors(previous.baseColors[i], next.baseColors[i], percent))
         }
         return Palette(colors = mixedColors.toList(), cycles = this.cycles)
     }
 
     fun cycle(timePassed: Int) {
         //it's important we copy the base values as each time we cycle it 'starts from 0'; it's not additive
-        colors = baseColors.toMutableList()
-        cycles
-                .filter { it.rate != 0 }
-                .forEach { cycle ->
+        baseColors.copyInto(colors)
+        cycles.forEach { cycle ->
+                    if(cycle.rate==0)
+                        return@forEach //Similar to break in a conventionnal loop
```

Comment:
I probably won't be able to run this locally until friday or this weekend.
Why not use filter? Is that not available on the intArray or is this somehow more performant?

Reply:
No worries, we are not in a rush !
I have to admit, it's my C training speaking there : like that we have to iterate only once on the array, instead of needing to iterate first to keep only the non zero values, and then iterate again to apply the color's algorithm.
Moreover, i'm pretty sure it prevents the creation of an intermediate array, containing only the non zero values.

Reply:
To my knowledge, the `.filter` _would_ create an intermediate list of the cycle and _would_ iterate a second time over those cycles that are not a 0 rate. However, I don't think that's much overhead given the cycle list is usually very small. (I may remember wrong but I'd guess 10 or less on average), so I'm not sure it's a large performance change.

That said, I think the return@forEach is like you said, the same as a 'break' in java, but I think what you'd need here is a 'continue'. I believe as written, this change would prevent any further cycles being evaluated after finding one cycle with a 0 rate. Since the cycles are not sorted, this could mean valid cycles are prevented.

If both my understandings are correct, I'd say it's worth keeping the filter for clarity since I don't think it will hurt performance without having so many cycles that another performance optimization would be better anyway.

Reply:
Yeah i'm sorry, i wanted to say 'continue' and not 'break', I goofed on that one. I can confirm that it indeed works like 'continue'.
I'll profile with and without that notation to confirm if it is useful or just an overly enthusiastic optimization !

Reply:
https://user-images.githubusercontent.com/10716685/175832101-a2c5c760-e5f0-4689-bd38-3bcde257dfd3.mp4
So I profiled both algorithms (with filter, and without filter+return@forEach) :
Sadly, my pc was dying while using the emulator + the profiler + the screencap, so there are not much fps here ...
With filter :
https://user-images.githubusercontent.com/10716685/175832145-49fbad02-9953-433f-8c26-2eb7f26a1e91.mp4

Without filter :
https://user-images.githubusercontent.com/10716685/175832153-d1966287-44ab-4ed2-a9ef-fed2cf6091ad.mp4

In conclusion it helps, but not by much.
I still have to optimize the creation of a 'Date()' object on each frame, but it's not really expensive haha